garrysmod/lua/includes/extensions/vector.lua

```diff
@@ -3,10 +3,262 @@ local meta = FindMetaTable( "Vector" )
+-----------------------------------------------------------]]
+function meta:Negate()
+	local x, y, z = self:Unpack()
+	self:SetUnpacked(-x, -y, -z)
+end
+
+--[[---------------------------------------------------------
+Returns a copy of the nagated vector
+-----------------------------------------------------------]]
+function meta:GetNegated(...)
+	local v = Vector(self)
+	v:Negate(...)
+	return v
+end
+
+--[[---------------------------------------------------------
+Convets the vector to nagated version of itself
+-----------------------------------------------------------]]
+function meta:Nudge(num, vec)
+	self:Add(vec:GetMul(num))
+end
+
+--[[---------------------------------------------------------
+Returns a copy of the nagated vector
+-----------------------------------------------------------]]
+function meta:GetNudge(...)
+	local v = Vector(self)
+	v:Nudge(...)
+	return v
+end
 
+--[[---------------------------------------------------------
+Convets the vector to resized version of itself
+-----------------------------------------------------------]]
+function meta:Resize(num)
+	self:Normalize()
+	self:Mul(num)
+end
+
+--[[---------------------------------------------------------
+Returns a copy of the resized vector
+-----------------------------------------------------------]]
+function meta:GetResized(...)
+	local v = Vector(self)
+	v:Resize(...)
+	return v
+end
+
+--[[---------------------------------------------------------
+Convets the vector to resized midpoint of itself
+-----------------------------------------------------------]]
+function meta:Midpoint(vec)
+	self:Add(vec)
+	self:Mul(0.5)
+end
+
+--[[---------------------------------------------------------
+Returns a copy of the midpoint vector
+-----------------------------------------------------------]]
+function meta:GetMidpoint(...)
+	local v = Vector(self)
+	v:Midpoint(...)
+	return v
+end
+
+--[[---------------------------------------------------------
+Returns a copy of the zero point vector
+-----------------------------------------------------------]]
+function meta:GetZero(...)
+	local v = Vector(self)
+	v:Zero(...)
+	return v
+end
+
+--[[---------------------------------------------------------
+Applies new basis to a vector by the three axises of an angle
+-----------------------------------------------------------]]
+function meta:Basis(ang)
+	local x = self:Dot(ang:Forward())
+	local y = self:Dot(ang:Right())
+	local z = self:Dot(ang:Up())
+	self:SetUnpacked(x, y, z)
+end
+
+--[[---------------------------------------------------------
+Applies new basis to a vector by the three axises of an angle
+-----------------------------------------------------------]]
+function meta:GetBasis(...)
+	local v = Vector(self)
+	v:Basis(...)
+	return v
+end
+
+--[[---------------------------------------------------------
+Projects the vector in to an axis defined by the second vector
+-----------------------------------------------------------]]
+function meta:Project(vec)

```

Comment:
It would be better to copy the original vector instead of using GetNormalized than setting it after finding the dot product.
```Lua
local v = Vector(self)
self:Normalize()
self:Mul(v:Dot(self))
```

Reply:
No, because it need Normalized of the argument as it defines the base X coordinate axis. You have to store the argumant in `self` rather.

Reply:
Yeah you're correct, mixed those up.